### DISCUSSION TITLE
Input type fields on input types don't seem to be updated after initialization

### DISCUSSION THREAD

**Discussion by madiganz** [profusion/sgqlc#199] [Q&A] [answered]
Don't know if this is expected or not, but when I have an input type as a field on another input type, I cannot set it after initialization
```python
class TestInput(sgqlc.types.Input):
    __schema__ = test_schema
    __field_names__ = ("field1", "field2")

    field1 = sgqlc.types.Field("FooInput", graphql_name="field1")

    field2 = sgqlc.types.Field(sgqlc.types.list_of("BarInput"), graphql_name="field2")

class FooInput(sgqlc.types.Input):
    __schema__ = test_schema
    __field_names__ = ("name")

    name = sgqlc.types.Field(String, graphql_name="name")

class BarInput(sgqlc.types.Input):
    __schema__ = test_schema
    __field_names__ = ("name")

    name = sgqlc.types.Field(String, graphql_name="name")
```

```python
# This works as expected
test = TestInput(
    field1=schema.FooInput(name: "test1"),
    field2=[schema.BarInput(name: "test2")]
)

test.field1 = schema.FooInput(name: "updated") # This does nothing if field1 was not set above
test.field2.append(schemaBarInput(name: "added")) # This does nothing
```

**Comment by barbieri** [comment, marked as answer]
In your examples, please note that `("name")` is the same as `"name"` (a string, not a tuple). You should use a tuple, then you need a trailing `,` as in `("name",)`.

The input types are meant to be used as a single, immutable construction as they will get the fields and convert to the GraphQL format, it's a difficult task to wrap every mutable type, so it's not done. In your second example `test.field2` is a simple `list` type, and mutating it won't update the GraphQL-converted data structure (it's only done in the constructor).

It's possible to be added, but an effort I can't do right now (patches are welcome 😀)

**Reply by madiganz** [reply]
Apologies, I realized my example is not correct, but you still answered my question, thanks!

**Reply by madiganz** [reply]
As a follow up, it is expected that setting `field1` in the constructor would cause the graphql to be an empty object? This causes failures as the `name` field is required (in my real example). I would think it would set the underlying graphql type to null

**Reply by barbieri** [reply]
not sure exactly what you mean, could you provide an example?

We can't convert an empty object to null automatically as they are different things.

**Reply by madiganz** [reply]
If we take a similar example to what I have above and do this:
```python
test = TestInput(
    field1=None,
    field2=[schema.BarInput(name: "test2")]
)
```
The graphql created for this will be
```graphql
input: {field1: {}, field2: {name: "test2"}}
```

**Reply by barbieri** [reply]
Hum, I need to test, it shouldn't produce an empty object, but rather a `null` in the output. Could you create a bug with an example so I remember to check that one?

**Reply by madiganz** [reply]
No problem - created a bug for it with an example. https://github.com/profusion/sgqlc/issues/200
